aws-amplify/aws-sdk-ios
# fix: Fix dependency declarations and linking for all targets

- Normalize deployment target & file inclusion for tests
- Remove hardcoded LD_RUNPATH_SEARCH_PATHS

*Issue #, if available:*

#2927
https://github.com/aws-amplify/aws-sdk-ios/issues/3036

**Description of changes:**

The previous project inconsistently linked frameworks in the Xcode project. SDK targets relied on "Dependency" build phases and `LD_RUNPATH_SEARCH_PATHS` build settings to track down and link build dependencies.

This change changes several build time settings:

- **Breaking change** AWSCognitoIdentityProviderASF no longer ships the prepackaged `libAWSCognitoIdentityProviderASFBinary.a` static library. AWSCognitoIdentityProviderASF is also now versioned the same as the rest of the SDK.
- Removes hardcoded `LD_RUNPATH_SEARCH_PATHS` from all targets. This appeared to be used prior to Xcode 12 to help resolve frameworks at linking time. However, our best information is that this attribute (which populates the linker's `rpath` argument), should be used to resolve paths at runtime, not build time. Apple's [guidance](https://developer.apple.com/videos/play/wwdc2018/415/) for resolving framework dependencies is to add them as explicit members of the "Dependency" build phase (which sets up an explicit dependency, but no linking), or as members of the "Link binary with libraries" build phase (which sets up an implicit dependency if that option is selected in the scheme editor, and also links as desired).
- For the most part, removes dependencies from the "Dependency" build phases in favor of adding them to the "Link binary with libraries" build phase. The latter is required to properly link referenced symbols; the former appears to be derivable from adding the "Link binaries" build phase. Some test targets retain a "Dependency" for `AWSAllTestsHost`, since that host must be built to act as a host target for the unit or integration test. In those targets, `AWSAllTestsHost` is *not* added to the "Link binary with libraries" build phase.
- For all test targets, links `AWSCore`, the SDK under test, and `AWSTestResources`. Some test targets may not actually use the `AWSTestResources` functionality, but it's innocuous to include it and makes for easier reference patterns for future SDK additions
- Removes unnecessary "Copy files" phases that inconsistently copied frameworks to build paths for unit tests only
- Builds Carthage and direct binaries with a script from https://github.com/Carthage/Carthage/issues/3019#issuecomment-665136323 . This script works around the error building Carthage binaries in Xcode 12

**Testing**

- [x] CircleCI builds, unit tests pass, Xcode 11
- [x] All unit tests pass, Xcode 11
- [x] All unit tests pass, Xcode 12
- [x] All integ tests build, Xcode 11
- [x] All integ tests build, Xcode 12
- [x] All SDK targets build locally in Xcode 11
- [x] All SDK targets build locally in Xcode 12
- [x] All SDK targets build via Carthage using new script, Xcode 11
- [x] All SDK targets build via Carthage using new script, Xcode 12
- [x] CocoaPods `pod lib lint`, Xcode 11
- [x] CocoaPods `pod lib lint`, Xcode 12
- [x] Amplify installs from pod, builds, Xcode 11
- [x] Amplify installs from pod, builds, Xcode 12
- Ensure sample apps consume & build as expected:
  - [x] Amplify, CocoaPods, Xcode 11 (@palpatim)
  - [x] Amplify, CocoaPods, Xcode 12 (@palpatim)
  - [x] SDK, CocoaPods, Xcode 11 (@wooj2)
  - [x] SDK, CocoaPods, Xcode 12 (@wooj2)
  - [x] SDK, Carthage, Xcode 11 (@royjit, and will build Carthage Xcode 11 archive for use in AppSync test below)
  - [x] SDK, Carthage, Xcode 12 (@royjit, and will build Carthage Xcode 11 archive for use in AppSync test below)

* From your SDK project workspace: `git fetch origin && git checkout -b palpatim/xcode-12-build-fix origin/palpatim/xcode-12-build-fix`
* Setup private spec repo for SDK - `./Scripts/setup_private_specs.sh 2.17.0 2.17.0`

* From your Amplify project workspace: `git fetch origin && git checkout -b palpatim/xcode-12-build origin/palpatim/xcode-12-build`
* Setup private spec repo for Amplify - `./Scripts/setup_private_specs.sh 1.2.1 1.2.1`

* Select Xcode: `sudo xcode-select --switch <path to /Applications/your Xcode version.app>`
* Verify Xcode selection: `xcode-select -p`

* Switch to your test app project workspace
* Sample project should have imports for all tested frameworks

* Carthage testing:
  - `./CircleciScripts/carthage-build.sh --platform iOS --no-skip-current --cache-builds | tee buildout.txt`
  - `rm -f CircleciScripts/.python-version && python3 CircleciScripts/create_carthage_archive.py && git checkout -- ./CircleciScripts/.python-version`
  - Copy the archive .zip file to your workstation, unzip it
  - Install required frameworks in your test project: "Link binary with libraries" build phase step
  - Quit Xcode
  - Clear derived data: `rm -rf $HOME/Library/Developer/Xcode/DerivedData`
  - Restart Xcode
  - Immediately build project before Indexing starts

* CocoaPods testing
  - Quit Xcode
  - Clear derived data: `rm -rf $HOME/Library/Developer/Xcode/DerivedData`
  - Clears all installed pods and cache: `find . -name Pods -o -name Podfile.lock | xargs rm -rf && pod cache clean --all`
  - Install pods: `root=$PWD ; find . -name Podfile | while read podPath ; do d=$( dirname $podPath ); cd $d && pod install --verbose && cd $root ; done`
    - Watch the output of the verbose pod install. You're looking to make sure that the pod installations are happening from your private spec repo, not trunk. E.g., you want to see a git clone like `/usr/bin/git clone file:///Users/...` not `/usr/bin/git clone https://github...`
  - Restart Xcode
  - Immediately build project before Indexing starts



---


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
